### PR TITLE
`AvlTree.insert` operation

### LINKED ISSUE #411: [400 SigmaUSD] `AvlTree.insert` operation
Create `AvlTreeData` struct and use in `Value::AvlTree(AvlTreeData)` and in `Literal::AvlTree(AvlTreeData)` and implement serialization (`SigmaSerializable` trait).
Scala reference implementation - https://github.com/ScorexFoundation/sigmastate-interpreter/blob/c863a9b1a82589e47b15f76f3affdb30a475e740/sigmastate/src/main/scala/sigmastate/AvlTreeData.scala#L42-L101

Create `savltree` modules for AvlTree methods definitions and evaluation similar to the existing `sbox` modules.
Add `AvlTree.insert` method in `savltree` module.
Scala reference implementation - https://github.com/ScorexFoundation/sigmastate-interpreter/blob/c863a9b1a82589e47b15f76f3affdb30a475e740/sigmastate/src/main/scala/sigmastate/types.scala#L2232-L2281

Use [`BatchAvlVerifier`](https://github.com/knizhnik/scorex_crypto_avltree/blob/36a0b47742d9d8f749ce95d2a794be88a03d13ec/src/batch_avl_verifier.rs#L8-L33) from https://crates.io/crates/scorex_crypto_avltree crate.

Use the existing `ADDigest` type for tree digest.

### PR DESCRIPTION
**Pull request by kettlebell**
Close #411

### CODE REVIEW

File: ergotree-interpreter/src/eval/savltree.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        })?;
+        Bytes::from(v.try_extract_into::<Vec<u8>>()?)
+    };
+
+    let starting_digest = Bytes::from(avl_tree_data.digest.0.to_vec());
+    let mut bv = BatchAVLVerifier::new(
+        &starting_digest,
+        &proof,
+        AVLTree::new(
+            |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
```
Comment (kettlebell) [inline ergotree-interpreter/src/eval/savltree.rs:86]: The Rust AVL library has this notion of `Resolver` that's not present in the scala implementation. I just copied what was used in the library example.

File: ergotree-interpreter/src/eval/savltree.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    use super::*;
+    #[test]
+    fn eval_avl_tree() {
+        // This example taken from `scorex_crypto_avltree` README
+        let mut prover = BatchAVLProver::new(
+            AVLTree::new(
+                |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
+                1,
+                None,
```
Comment (kettlebell) [inline ergotree-interpreter/src/eval/savltree.rs:145]: I'm not sure how this parameter is used for AVL operations. Setting it to None here and for `AvlTreeData::value_length_opt` allowed the test to pass.

**Comment by greenhat** [comment]
`scorex_crypto_avltree` addition broke `ergo-lib-wasm` build by pulling new `rand`. Please update `rand` version in all our crates to `0.8.4`(or .3) and 
update `getrandom = {version = "0.1", features = ["wasm-bindgen"]}` in `ergo-lib-wasm` to `getrandom = {version = "0.2.3", features = ["js"]}`.
`

**Review by greenhat — CHANGES_REQUESTED** [review]
Great job figuring out AVL trees! Looking great! Check minor comments.

**Comment by greenhat** [comment]
> > `scorex_crypto_avltree` addition broke `ergo-lib-wasm` build by pulling new `rand`. Please update `rand` version in all our crates to `0.8.4`(or .3) and update `getrandom = {version = "0.1", features = ["wasm-bindgen"]}` in `ergo-lib-wasm` to `getrandom = {version = "0.2.3", features = ["js"]}`. `
> 
> I've updated the dependencies in my latest commit, but CI is totally broken as you can see. Each test refers to errors in ergotree-ir/src/mir/constant.rs, but the code doesn't match the commit. Not sure what's going on.
> 
> Also please note that I had to update more than just `rand` and `getrandom`. Details in commit message.

Judging by compiler error I believe you need to rebase this branch onto fresh develop, after I merged SubstConstants PR. For some reason Github UI did not pick up new develop branch.

File: ergotree-ir/src/mir/constant.rs [hunk 3]
```diff
@@ -53,6 +54,8 @@ pub enum Literal {
     GroupElement(Box<EcPoint>),
     /// Ergo box ID
     CBox(IrBoxId),
+    /// AVL tree
+    AvlTree(AvlTreeData),
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-ir/src/mir/constant.rs]: Done.

File: ergotree-ir/src/mir/value.rs [hunk 4]
```diff
@@ -150,7 +151,7 @@ pub enum Value {
     /// Box
     CBox(IrBoxId),
     /// AVL tree
-    AvlTree,
+    AvlTree(AvlTreeData),
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-ir/src/mir/value.rs]: Done.

File: ergotree-ir/src/types/savltree.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            &INSERT_METHOD_DESC,
+        ]
+    ;
+}
+
+lazy_static! {
+    static ref INSERT_METHOD_DESC: SMethodDesc = SMethodDesc {
+        method_id: INSERT_METHOD_ID,
+        name: "insert",
+        tpe: SFunc {
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-ir/src/types/savltree.rs:32]: Thanks.

File: ergotree-ir/src/mir/avl_tree_data.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    /// If non-empty, all the values under the tree are of the same length
+    pub value_length_opt: Option<u32>,
+}
+
+impl SigmaSerializable for AvlTreeData {
+    fn sigma_serialize<W: SigmaByteWrite>(&self, w: &mut W) -> SigmaSerializeResult {
+        self.digest.sigma_serialize(w)?;
+        w.put_u8(self.tree_flags.0)?;
+        w.put_u32(self.key_length)?;
+        if let Some(l) = self.value_length_opt {
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-ir/src/mir/avl_tree_data.rs]: There's existing impl for `Option<Box<T>>` so I went with that (trying to implement it for `Option<T>` lead to compiler complaints).

File: ergotree-ir/src/mir/avl_tree_data.rs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        } else {
+            w.put_u8(0)?;
+        }
+        Ok(())
+    }
+    fn sigma_parse<R: SigmaByteRead>(r: &mut R) -> Result<Self, SigmaParsingError> {
+        let digest = ADDigest::sigma_parse(r)?;
+        let tree_flags = AvlTreeFlags::parse(r.get_u8()?);
+        let key_length = r.get_u32()?;
+        let is_some = {
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-ir/src/mir/avl_tree_data.rs]: Done.

File: ergotree-interpreter/src/eval/savltree.rs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+use scorex_crypto_avltree::batch_node::Node;
+use scorex_crypto_avltree::batch_node::NodeHeader;
+use scorex_crypto_avltree::operation::KeyValue;
+use scorex_crypto_avltree::operation::Operation;
+
+use super::EvalError;
+use super::EvalFn;
+
+pub(crate) static INSERT_EVAL_FN: EvalFn = |_env, _ctx, obj, args| {
+    let mut avl_tree_data = match obj {
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-interpreter/src/eval/savltree.rs]: Done.

File: ergotree-interpreter/src/eval/savltree.rs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        Value::AvlTree(a) => Ok(a),
+        v => Err(EvalError::UnexpectedValue(format!(
+            "AvlTree.insert: expected object of Value::AvlTree got: {0:?}",
+            v
+        ))),
+    }?;
+
+    if !avl_tree_data.tree_flags.insert_allowed() {
+        return Err(EvalError::Misc("AvlTree: Insertions not allowed".into()));
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-interpreter/src/eval/savltree.rs]: I went with `EvalError::AvlTree(String)`.

File: ergotree-interpreter/src/eval/savltree.rs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    let entries = {
+        let v = args
+            .get(0)
+            .cloned()
+            .ok_or_else(|| EvalError::NotFound("AvlTree: eval is missing first arg".to_string()))?;
+        match v {
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-interpreter/src/eval/savltree.rs]: Thanks, saved a lot of lines. Need to remember to use this trait more!

File: ergotree-ir/src/serialization/data.rs [hunk 11]
```diff
@@ -37,6 +37,7 @@ impl DataSerializer {
             Literal::GroupElement(ecp) => ecp.sigma_serialize(w)?,
             Literal::SigmaProp(s) => s.value().sigma_serialize(w)?,
             Literal::CBox(_) => return Err(SigmaSerializationError::NotImplementedYet("Box")),
+            Literal::AvlTree(a) => a.sigma_serialize(w)?,
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-ir/src/serialization/data.rs:42]: Done.

**Comment by kettlebell** [comment]
> > > `scorex_crypto_avltree` addition broke `ergo-lib-wasm` build by pulling new `rand`. Please update `rand` version in all our crates to `0.8.4`(or .3) and update `getrandom = {version = "0.1", features = ["wasm-bindgen"]}` in `ergo-lib-wasm` to `getrandom = {version = "0.2.3", features = ["js"]}`. `
> > 
> > 
> > I've updated the dependencies in my latest commit, but CI is totally broken as you can see. Each test refers to errors in ergotree-ir/src/mir/constant.rs, but the code doesn't match the commit. Not sure what's going on.
> > Also please note that I had to update more than just `rand` and `getrandom`. Details in commit message.
> 
> Judging by compiler error I believe you need to rebase this branch onto fresh develop, after I merged SubstConstants PR. For some reason Github UI did not pick up new develop branch.

Done, plus addressed all your concerns (with one little exception for the unresolved conversation above).

**Review by greenhat — APPROVED** [review]
Looking great! Thank you!

**Pull request merged**